# Projeto de Teste SQL

O principal objetivo do estudo é entender o mercado atual de livros. Com a mudança na rotina da população, devido ao coronavírus, muitas pessoas passaram a ficar em casa lendo livros, em vez de fora de casa. O que criou uma nova oportunidade de negócio. Para avaliar esse novo mercado foi analisado um banco de dados de um dos serviços concorrentes nesse mercado. Ele contém dados sobre livros, editoras, autores e classificação de clientes e avaliação de livros. Com essas informações será possível gerar um proposição válida para um novo produto.

# Estudando as tarefas

In [1]:
# importando bibliotecas
import pandas as pd
from matplotlib import pyplot as plt
import numpy as np
import seaborn as sns
from math import factorial
from scipy import stats as st
import math as mt
import plotly.express as px
from datetime import datetime
import requests
from bs4 import BeautifulSoup
import re
from sqlalchemy import create_engine

In [2]:
db_config = {'user': 'practicum_student', # nome de usuário
'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # senha
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # porta de conexão
'db': 'data-analyst-final-project-db'} # o nome do banco de dados
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
db_config['pwd'],
db_config['host'],
db_config['port'],
db_config['db'])
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [3]:
query = 'SELECT * FROM books;'
book = pd.io.sql.read_sql(query, con = engine)

In [4]:
book.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


In [5]:
book.head(10)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

In [6]:
book_duplicated = book[book.duplicated()]
print(book_duplicated)

Empty DataFrame
Columns: [book_id, author_id, title, num_pages, publication_date, publisher_id]
Index: []


- a data de publicação foi convertida em datetime
- as demais colunas estão com o tipo certo
- não existem valores ausentes
- não existem valores duplicados

In [7]:
query2 = 'SELECT * FROM authors;'
author = pd.io.sql.read_sql(query2, con = engine)

In [8]:
author.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


In [9]:
author.head(10)

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

In [10]:
author_duplicated = author[author.duplicated()]
print(author_duplicated)

Empty DataFrame
Columns: [author_id, author]
Index: []


- os dados estão com o tipo correto
- não existem valores ausentes
- não existem valores duplicados

In [11]:
query3 = 'SELECT * FROM publishers;'
publisher = pd.io.sql.read_sql(query3, con = engine)

In [12]:
publisher.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


In [13]:
publisher.head(10)

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

In [14]:
publisher_duplicated = publisher[publisher.duplicated()]
print(publisher_duplicated)

Empty DataFrame
Columns: [publisher_id, publisher]
Index: []


- os dados estão com os tipos corretos
- não existem valores ausentes
- não existem valores duplicados

In [15]:
query4 = 'SELECT * FROM ratings;'
rating = pd.io.sql.read_sql(query4, con = engine)

In [16]:
rating.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


In [17]:
rating.head(10)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

In [18]:
rating_duplicated = rating[rating.duplicated()]
print(rating_duplicated)

Empty DataFrame
Columns: [rating_id, book_id, username, rating]
Index: []


- os dados estão com os tipos corretos
- não existem valores ausentes
- não existem valores duplciados

In [19]:
query5 = 'SELECT * FROM reviews;'
review = pd.io.sql.read_sql(query5, con = engine)

In [20]:
review.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


In [21]:
review.head(10)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

In [22]:
review_duplicated = review[review.duplicated()]
print(review_duplicated)

Empty DataFrame
Columns: [review_id, book_id, username, text]
Index: []


- os dados estão com os tipos corretos
- não existem valores ausentes
- não existem valores duplciados

O número total de livros lançados depois de 01/01/2020 foi 463

In [23]:
query9 = """
SELECT COUNT(*) 
FROM books 
WHERE publication_date > '2000-01-01';
"""
books = pd.read_sql(query9, con=engine)
print('Número de livros:', books)

Número de livros:    count
0    819


A quantidade de livros lançados depois de 01/01/2020 foi 819 livros

In [24]:
query10 = """
SELECT COUNT(rating)
FROM ratings;
"""

countrating = pd.read_sql(query10, con=engine)
print('Número de avaliações:', countrating)

Número de avaliações:    count
0   6456


A quantidade de avaliações é 6456

In [25]:
query20 = """SELECT b.title, AVG(r.rating) AS average_rating
FROM books 
LEFT JOIN ratings r ON b.book_id = r.book_id
GROUP BY b.title;
"""

In [34]:
query11 = """
SELECT books.title, subquery.avg_rating, subquery.review_count
FROM (
    SELECT reviews.book_id as book_id, 
           COUNT(DISTINCT reviews.review_id) AS review_count, 
           AVG(ratings.rating) AS avg_rating
    FROM reviews 
    INNER JOIN ratings ON ratings.book_id = reviews.book_id
    GROUP BY reviews.book_id
) AS subquery
INNER JOIN books ON subquery.book_id = books.book_id
ORDER BY subquery.review_count DESC, subquery.avg_rating DESC
LIMIT 10
"""
avgrating = pd.read_sql(query11, con=engine)
avgrating

title  avg_rating  review_count
0                            Twilight (Twilight  #1)    3.662500             7
1  Harry Potter and the Prisoner of Azkaban (Harr...    4.414634             6
2  Harry Potter and the Chamber of Secrets (Harry...    4.287500             6
3                                     The Book Thief    4.264151             6
4                                   The Glass Castle    4.206897             6
5                          Outlander (Outlander  #1)    4.125000             6
6                The Hobbit  or There and Back Again    4.125000             6
7  The Curious Incident of the Dog in the Night-Time    4.081081             6
8  The Lightning Thief (Percy Jackson and the Oly...    4.080645             6
9                                Water for Elephants    3.977273             6

A tabela acima mostra a classificação média dos livros em ordem de livros com mais avaliações

In [36]:
query12 = """
SELECT COUNT(books.book_id) AS count, publishers.publisher AS publisher
FROM books
INNER JOIN publishers ON publishers.publisher_id = books.publisher_id
WHERE books.num_pages > 50
GROUP BY publishers.publisher
ORDER BY count DESC
LIMIT 5
"""
pagesmax = pd.read_sql(query12, con=engine)

pagesmax

count                 publisher
0     42             Penguin Books
1     31                   Vintage
2     25  Grand Central Publishing
3     24          Penguin Classics
4     19                    Bantam

Na tabela acima, mostramos os 5 livros com mais publicações de livros com mais de 50 páginas.

In [93]:
query40 = """
WITH AuthorRatings AS (
    SELECT b.author_id, AVG(r.rating) AS average_rating, COUNT(r.rating) AS rating_count
    FROM ratings r
    JOIN books b ON b.book_id = r.book_id
    GROUP BY b.author_id
    HAVING COUNT(r.rating) > 50  
)
SELECT ar.average_rating, a.author AS author_name
FROM AuthorRatings ar
JOIN authors a ON ar.author_id = a.author_id
ORDER BY ar.average_rating DESC
LIMIT 1;
"""

ratingauthor = pd.read_sql(query40, con=engine)

ratingauthor

average_rating                 author_name
0        4.288462  J.K. Rowling/Mary GrandPré

Na tabela acima, mostramos a escritora com a maior classificação média, com mais de 50 avaliações. 

In [41]:
query55 = """
SELECT AVG(subquery2.review_cnt) AS avg_review_cnt
FROM (
    SELECT
        COUNT(reviews.review_id) AS review_cnt, subquery1.username
    FROM (
        SELECT username, COUNT(rating_id) AS rating_cnt
        FROM ratings
        GROUP BY username
        HAVING COUNT(rating_id) > 50
    ) AS subquery1
    INNER JOIN reviews ON reviews.username = subquery1.username
    GROUP BY subquery1.username
) AS subquery2
"""

meanrating = pd.read_sql(query55, con=engine)
meanrating

avg_review_cnt
0       24.333333

Na tabela acima, mostramos a média das avaliações entre os usuários que classificaram mais do que 50 livros.

# Conclusão

- com os dados é possível perceber que realmente é um mercado que vale a pena investir
- a quantidade de livros publicados foi bem grande
- existe um grande número de usuários que avaliam os livros, o que mostra que eles estão motivados a continuar a ler e mostrar sua opnião para próximos livros e estilos de livros que mais gostam
- concluindo, vale a pena investir em aplciativos para amantes de livros, seja ele apenas um lugar para os usuários mostrarem suas opniões, como também um lugar para mostrar livros com estilos parecidos com o que eles leram, opniões de outros usuários, onde comprar, entre outras coisas.